spec/java/util/stream/DoubleStream.main.lsl

```diff
@@ -0,0 +1,1266 @@
+
+            result = new DoubleStreamAutomaton(state = Initialized,
+                storage = newStorage,
+                length = newLength,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+
+        this.linkedOrConsumed = true;
+    }
+
+
+    @Phantom proc _takeWhile_loop (i: int, takeLength: int, predicate: DoublePredicate): void
+    {
+        if (action CALL(predicate, [this.storage[i]]))
+            takeLength += 1;
+        else
+            action LOOP_BREAK();
+    }
+
+
+    @Phantom proc _copy_takeWhile_loop (i: int, j: int, newStorage: array<double>): void
+    {
+        newStorage[j] = this.storage[i];
+        j += 1;
+    }
+
+
+    fun *.asLongStream (@target self: DoubleStream): LongStream
+    {
+        if (this.length == 0)
+        {
+            val emptyArray: array<long> = action ARRAY_NEW("long", 0);
+            result = new LongStreamAutomaton(state = Initialized,
+                storage = emptyArray,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        else
+        {
+            val newStorage: array<long> = action ARRAY_NEW("long", this.length);
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, this.length, +1,
+                _convertDoubleStorageToLong_loop(i, newStorage)
+            );
+
+            result = new LongStreamAutomaton(state = Initialized,
+                storage = newStorage,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc _convertDoubleStorageToLong_loop(i: int, newStorage: array<long>): void
+    {
+        newStorage[i] = this.storage[i] as long;
+    }
+
+
+    fun *.asIntStream (@target self: DoubleStream): IntStream
+    {
+        if (this.length == 0)
+        {
+            val emptyArray: array<int> = action ARRAY_NEW("int", 0);
+            result = new IntStreamAutomaton(state = Initialized,
+                storage = emptyArray,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        else
+        {
+            val newStorage: array<int> = action ARRAY_NEW("int", this.length);
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, this.length, +1,
+                _convertDoubleStorageToInt_loop(i, newStorage)
+            );
+
+            result = new IntStreamAutomaton(state = Initialized,
+                storage = newStorage,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc _convertDoubleStorageToInt_loop(i: int, newStorage: array<int>): void
+    {
+        newStorage[i] = this.storage[i] as int;
+    }
+
+
+    fun *.sum (@target self: DoubleStream): double
+    {
+        result = 0;
+
+        if (this.length != 0)
```

Comment:
Why not move this check into `_sum()` itself?

Reply:
Fixed